File: .UNSAFE-ALL-BINDINGS-DELETED-FINAL-20260125/bindings/php/src/Proven/SafeHex.php
<?php
// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2025 Hyperpolymath

declare(strict_types=1);

namespace Proven;

/**
 * Safe hexadecimal encoding/decoding operations.
 *
 * Provides constant-time operations where security-relevant and
 * safe handling of invalid input.
 */
class SafeHex
{
    /** @var string Lowercase hex alphabet */
    private const HEX_LOWER = '0123456789abcdef';

    /** @var string Uppercase hex alphabet */
    private const HEX_UPPER = '0123456789ABCDEF';

    /**
     * Encode binary data to lowercase hexadecimal.
     *
     * @param string $data The binary data to encode
     * @return string Lowercase hex string
     */
    public static function encode(string $data): string
    {
        return bin2hex($data);
    }

    /**
     * Encode binary data to uppercase hexadecimal.
     *
     * @param string $data The binary data to encode
     * @return string Uppercase hex string
     */
    public static function encodeUpper(string $data): string
    {
        return strtoupper(bin2hex($data));
    }

    /**
     * Decode hexadecimal string to binary data.
     *
     * @param string $hex The hex string to decode
     * @return string|null Decoded binary data or null if invalid
     */
    public static function decode(string $hex): ?string
    {
        // Remove optional 0x prefix
        if (str_starts_with($hex, '0x') || str_starts_with($hex, '0X')) {
            $hex = substr($hex, 2);
        }

        // Check for valid hex string (even length, hex characters only)
        if (strlen($hex) % 2 !== 0) {
            return null;
        }

        if (!ctype_xdigit($hex) && $hex !== '') {
            return null;
        }

        if ($hex === '') {
            return '';
        }

        $result = @hex2bin($hex);
        return $result === false ? null : $result;
    }

    /**
     * Decode hexadecimal with a fallback value for invalid input.
     *
     * @param string $hex The hex string to decode
     * @param string $fallback Value to return if invalid
     * @return string Decoded data or fallback
     */
    public static function decodeOr(string $hex, string $fallback): string
    {
        return self::decode($hex) ?? $fallback;
    }

    /**
     * Compare two hex strings in constant time.
     *
     * This prevents timing attacks when comparing secrets like API keys,
     * tokens, or other sensitive hex-encoded data.
     *
     * @param string $hexA First hex string
     * @param string $hexB Second hex string
     * @return bool True if equal (case-insensitive)
     */
    public static function constantTimeEqual(string $hexA, string $hexB): bool
    {
        // Normalize both to lowercase for comparison
        $normalizedA = strtolower($hexA);
        $normalizedB = strtolower($hexB);

        // Remove 0x prefix if present
        if (str_starts_with($normalizedA, '0x')) {
            $normalizedA = substr($normalizedA, 2);
        }
        if (str_starts_with($normalizedB, '0x')) {
            $normalizedB = substr($normalizedB, 2);
        }

        return hash_equals($normalizedA, $normalizedB);
    }

    /**
     * Check if a string is valid hexadecimal.
     *
     * @param string $hex The string to validate
     * @param bool $allowPrefix Whether to allow 0x prefix
     * @return bool True if valid hex
     */
    public static function isValid(string $hex, bool $allowPrefix = true): bool
    {
        $toCheck = $hex;

        if ($allowPrefix && (str_starts_with($hex, '0x') || str_starts_with($hex, '0X'))) {
            $toCheck = substr($hex, 2);
        }

        // Empty string after removing prefix is not valid hex
        if ($toCheck === '') {
            return false;
        }

        return ctype_xdigit($toCheck);
    }

    /**
     * Check if a string is valid hexadecimal with even length (can be decoded).
     *
     * @param string $hex The string to validate
     * @param bool $allowPrefix Whether to allow 0x prefix
     * @return bool True if valid and decodable
     */
    public static function isDecodable(string $hex, bool $allowPrefix = true): bool
    {
        $toCheck = $hex;

        if ($allowPrefix && (str_starts_with($hex, '0x') || str_starts_with($hex, '0X'))) {
            $toCheck = substr($hex, 2);
        }

        if ($toCheck === '') {
            return true; // Empty hex decodes to empty string
        }

        return ctype_xdigit($toCheck) && strlen($toCheck) % 2 === 0;
    }

    /**
     * Normalize a hex string (lowercase, no prefix).
     *
     * @param string $hex The hex string to normalize
     * @return string|null Normalized hex or null if invalid
     */
    public static function normalize(string $hex): ?string
    {
        $toNormalize = $hex;

        if (str_starts_with($hex, '0x') || str_starts_with($hex, '0X')) {
            $toNormalize = substr($hex, 2);
        }

        if ($toNormalize !== '' && !ctype_xdigit($toNormalize)) {
            return null;
        }

        return strtolower($toNormalize);
    }

    /**
     * Add 0x prefix to hex string if not present.
     *
     * @param string $hex The hex string
     * @return string|null Hex with 0x prefix or null if invalid
     */
    public static function withPrefix(string $hex): ?string
    {
        $normalized = self::normalize($hex);
        if ($normalized === null) {
            return null;
        }
        return '0x' . $normalized;
    }

    /**
     * Remove 0x prefix from hex string if present.
     *
     * @param string $hex The hex string
     * @return string|null Hex without prefix or null if invalid
     */
    public static function withoutPrefix(string $hex): ?string
    {
        return self::normalize($hex);
    }

    /**
     * Encode a single byte to two hex characters.
     *
     * @param int $byte Byte value (0-255)
     * @return string Two hex characters
     * @throws \InvalidArgumentException If byte is out of range
     */
    public static function encodeByte(int $byte): string
    {
        if ($byte < 0 || $byte > 255) {
            throw new \InvalidArgumentException('Byte must be in range 0-255');
        }
        return sprintf('%02x', $byte);
    }

    /**
     * Decode two hex characters to a byte value.
     *
     * @param string $hex Two hex characters
     * @return int|null Byte value (0-255) or null if invalid
     */
    public static function decodeByte(string $hex): ?int
    {
        if (strlen($hex) !== 2 || !ctype_xdigit($hex)) {
            return null;
        }
        return (int)hexdec($hex);
    }

    /**
     * Encode an integer to hexadecimal.
     *
     * @param int $value The integer to encode
     * @param int $minWidth Minimum width (will be zero-padded)
     * @param bool $prefix Whether to include 0x prefix
     * @return string Hex representation
     */
    public static function encodeInt(int $value, int $minWidth = 0, bool $prefix = false): string
    {
        $hex = dechex($value);

        if ($minWidth > 0 && strlen($hex) < $minWidth) {
            $hex = str_pad($hex, $minWidth, '0', STR_PAD_LEFT);
        }

        return $prefix ? '0x' . $hex : $hex;
    }

    /**
     * Decode hexadecimal to an integer.
     *
     * @param string $hex The hex string
     * @return int|null The integer value or null if invalid
     */
    public static function decodeInt(string $hex): ?int
    {
        $normalized = self::normalize($hex);
        if ($normalized === null || $normalized === '') {
            return null;
        }

        // Prevent overflow on 64-bit systems (max 16 hex chars = 8 bytes)
        if (strlen($normalized) > 16) {
            return null;
        }

        return (int)hexdec($normalized);
    }

    /**
     * Format hex string with separators for readability.
     *
     * @param string $hex The hex string
     * @param string $separator Separator character (default space)
     * @param int $groupSize Number of chars per group (default 2)
     * @return string|null Formatted hex or null if invalid
     */
    public static function format(string $hex, string $separator = ' ', int $groupSize = 2): ?string
    {
        $normalized = self::normalize($hex);
        if ($normalized === null) {
            return null;
        }

        if ($normalized === '') {
            return '';
        }

        if ($groupSize <= 0) {
            return $normalized;
        }

        return implode($separator, str_split($normalized, $groupSize));
    }

    /**
     * Remove all whitespace and separators from hex string.
     *
     * @param string $hex The hex string (may contain spaces, colons, etc.)
     * @return string|null Clean hex or null if invalid after cleaning
     */
    public static function clean(string $hex): ?string
    {
        // Remove common separators: space, colon, dash, dot
        $cleaned = preg_replace('/[\s:\-.]/', '', $hex);
        if ($cleaned === null) {
            return null;
        }

        return self::normalize($cleaned);
    }

    /**
     * Generate a random hex string.
     *
     * @param int $byteLength Number of random bytes (output will be 2x length)
     * @return string Random hex string
     * @throws \Exception If secure random generation fails
     */
    public static function random(int $byteLength): string
    {
        if ($byteLength < 0) {
            throw new \InvalidArgumentException('Byte length must be non-negative');
        }
        if ($byteLength === 0) {
            return '';
        }
        return bin2hex(random_bytes($byteLength));
    }

    /**
     * XOR two hex strings of equal length.
     *
     * @param string $hexA First hex string
     * @param string $hexB Second hex string
     * @return string|null XOR result or null if lengths differ or invalid
     */
    public static function xor(string $hexA, string $hexB): ?string
    {
        $bytesA = self::decode($hexA);
        $bytesB = self::decode($hexB);

        if ($bytesA === null || $bytesB === null) {
            return null;
        }

        if (strlen($bytesA) !== strlen($bytesB)) {
            return null;
        }

        $result = $bytesA ^ $bytesB;
        return bin2hex($result);
    }

    /**
     * Check if two hex strings are equal (case-insensitive, ignores prefix).
     *
     * This is NOT constant-time - use constantTimeEqual for secrets.
     *
     * @param string $hexA First hex string
     * @param string $hexB Second hex string
     * @return bool True if equal
     */
    public static function equals(string $hexA, string $hexB): bool
    {
        $normalizedA = self::normalize($hexA);
        $normalizedB = self::normalize($hexB);

        if ($normalizedA === null || $normalizedB === null) {
            return false;
        }

        return $normalizedA === $normalizedB;
    }

    /**
     * Reverse the bytes in a hex string (for endianness conversion).
     *
     * @param string $hex The hex string
     * @return string|null Reversed hex or null if invalid/odd length
     */
    public static function reverseBytes(string $hex): ?string
    {
        $normalized = self::normalize($hex);
        if ($normalized === null || strlen($normalized) % 2 !== 0) {
            return null;
        }

        if ($normalized === '') {
            return '';
        }

        $bytes = str_split($normalized, 2);
        return implode('', array_reverse($bytes));
    }

    /**
     * Pad a hex string to a minimum length.
     *
     * @param string $hex The hex string
     * @param int $minLength Minimum length (in hex chars)
     * @param bool $padLeft Whether to pad on left (true) or right (false)
     * @return string|null Padded hex or null if invalid
     */
    public static function pad(string $hex, int $minLength, bool $padLeft = true): ?string
    {
        $normalized = self::normalize($hex);
        if ($normalized === null) {
            return null;
        }

        if (strlen($normalized) >= $minLength) {
            return $normalized;
        }

        $padType = $padLeft ? STR_PAD_LEFT : STR_PAD_RIGHT;
        return str_pad($normalized, $minLength, '0', $padType);
    }

    /**
     * Truncate a hex string to a maximum length.
     *
     * @param string $hex The hex string
     * @param int $maxLength Maximum length (in hex chars)
     * @param bool $fromLeft Whether to keep left (true) or right (false) portion
     * @return string|null Truncated hex or null if invalid
     */
    public static function truncate(string $hex, int $maxLength, bool $fromLeft = true): ?string
    {
        $normalized = self::normalize($hex);
        if ($normalized === null) {
            return null;
        }

        if (strlen($normalized) <= $maxLength) {
            return $normalized;
        }

        if ($fromLeft) {
            return substr($normalized, 0, $maxLength);
        }
        return substr($normalized, -$maxLength);
    }
}

File: bindings/java/src/main/java/com/proven/SafeHex.java
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2025 Hyperpolymath
package com.proven;

import java.nio.charset.StandardCharsets;

/**
 * Safe hexadecimal encoding and decoding.
 * Provides constant-time comparison for security-sensitive operations.
 * Calls native verified code via JNI when available.
 */
public final class SafeHex {
    private SafeHex() {}

    private static final char[] HEX_LOWER = "0123456789abcdef".toCharArray();
    private static final char[] HEX_UPPER = "0123456789ABCDEF".toCharArray();

    /**
     * Encode bytes to lowercase hex string.
     */
    public static String encode(byte[] bytes) {
        if (bytes == null) return "";
        if (ProvenNative.isNativeLoaded()) {
            int[] status = new int[1];
            byte[] result = ProvenNative.nativeHexEncode(bytes, false, status);
            if (ProvenStatus.fromCode(status[0]).isOk() && result != null) {
                return new String(result, StandardCharsets.UTF_8);
            }
        }
        return encodePure(bytes, false);
    }

    /**
     * Encode bytes to uppercase hex string.
     */
    public static String encodeUpper(byte[] bytes) {
        if (bytes == null) return "";
        if (ProvenNative.isNativeLoaded()) {
            int[] status = new int[1];
            byte[] result = ProvenNative.nativeHexEncode(bytes, true, status);
            if (ProvenStatus.fromCode(status[0]).isOk() && result != null) {
                return new String(result, StandardCharsets.UTF_8);
            }
        }
        return encodePure(bytes, true);
    }

    /**
     * Decode hex string to bytes.
     */
    public static ProvenResult<byte[]> decode(String hex) {
        if (hex == null || hex.isEmpty()) {
            return ProvenResult.ok(new byte[0]);
        }
        if (hex.length() % 2 != 0) {
            return ProvenResult.err(ProvenStatus.ERR_INVALID_ARGUMENT, "Hex string has odd length");
        }
        if (ProvenNative.isNativeLoaded()) {
            int[] status = new int[1];
            byte[] result = ProvenNative.nativeHexDecode(
                hex.getBytes(StandardCharsets.UTF_8), status);
            ProvenStatus s = ProvenStatus.fromCode(status[0]);
            return s.isOk() && result != null ? ProvenResult.ok(result) : ProvenResult.err(s);
        }
        return decodePure(hex);
    }

    /**
     * Check if character is valid hex digit.
     */
    public static boolean isValidChar(char c) {
        return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
    }

    /**
     * Check if string contains only valid hex characters.
     */
    public static boolean isValid(String hex) {
        if (hex == null) return false;
        if (ProvenNative.isNativeLoaded()) {
            return ProvenNative.nativeHexIsValid(hex.getBytes(StandardCharsets.UTF_8));
        }
        for (int i = 0; i < hex.length(); i++) {
            if (!isValidChar(hex.charAt(i))) return false;
        }
        return true;
    }

    /**
     * Check if string is valid hex bytes (even length, valid chars).
     */
    public static boolean isValidBytes(String hex) {
        return hex != null && hex.length() % 2 == 0 && isValid(hex);
    }

    /**
     * Constant-time hex string comparison.
     * Use for comparing security-sensitive values like API keys, tokens, and hashes.
     */
    public static boolean constantTimeEquals(String a, String b) {
        if (a == null || b == null) return false;
        if (a.length() != b.length()) return false;
        if (ProvenNative.isNativeLoaded()) {
            return ProvenNative.nativeHexConstantTimeEq(
                a.getBytes(StandardCharsets.UTF_8),
                b.getBytes(StandardCharsets.UTF_8));
        }
        return constantTimeEqualsPure(a, b);
    }

    /**
     * Constant-time byte array comparison.
     */
    public static boolean constantTimeEquals(byte[] a, byte[] b) {
        if (a == null || b == null) return false;
        if (a.length != b.length) return false;
        int result = 0;
        for (int i = 0; i < a.length; i++) {
            result |= a[i] ^ b[i];
        }
        return result == 0;
    }

    /**
     * Format hex with spaces between bytes.
     */
    public static String formatSpaced(String hex) {
        if (hex == null || hex.length() < 2) return hex;
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < hex.length(); i += 2) {
            if (i > 0) sb.append(' ');
            sb.append(hex, i, Math.min(i + 2, hex.length()));
        }
        return sb.toString();
    }

    /**
     * Format hex with colons between bytes (like MAC address).
     */
    public static String formatColons(String hex) {
        if (hex == null || hex.length() < 2) return hex;
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < hex.length(); i += 2) {
            if (i > 0) sb.append(':');
            sb.append(hex, i, Math.min(i + 2, hex.length()));
        }
        return sb.toString();
    }

    /**
     * Convert integer to hex string with optional minimum width.
     */
    public static String fromInt(long value, int minWidth, boolean uppercase) {
        String hex = Long.toHexString(value);
        if (uppercase) hex = hex.toUpperCase();
        if (hex.length() < minWidth) {
            hex = "0".repeat(minWidth - hex.length()) + hex;
        }
        return hex;
    }

    /**
     * Parse hex string to integer.
     */
    public static ProvenResult<Long> toInt(String hex) {
        if (hex == null || hex.isEmpty()) {
            return ProvenResult.err(ProvenStatus.ERR_EMPTY_INPUT, "Empty hex string");
        }

        String clean = hex;
        if (clean.startsWith("0x") || clean.startsWith("0X")) {
            clean = clean.substring(2);
        }

        if (!isValid(clean)) {
            return ProvenResult.err(ProvenStatus.ERR_INVALID_ARGUMENT, "Invalid hex characters");
        }

        try {
            return ProvenResult.ok(Long.parseUnsignedLong(clean, 16));
        } catch (NumberFormatException e) {
            return ProvenResult.err(ProvenStatus.ERR_OVERFLOW, "Hex value too large");
        }
    }

    /**
     * Strip 0x prefix if present.
     */
    public static String stripPrefix(String hex) {
        if (hex == null) return null;
        if (hex.startsWith("0x") || hex.startsWith("0X")) {
            return hex.substring(2);
        }
        return hex;
    }

    /**
     * Calculate required buffer size for encoding.
     */
    public static int encodeSize(int bytesLen) {
        return bytesLen * 2;
    }

    /**
     * Calculate required buffer size for decoding.
     */
    public static int decodeSize(int hexLen) {
        return hexLen / 2;
    }

    // Pure Java implementations

    private static String encodePure(byte[] bytes, boolean uppercase) {
        char[] table = uppercase ? HEX_UPPER : HEX_LOWER;
        char[] result = new char[bytes.length * 2];
        for (int i = 0; i < bytes.length; i++) {
            int b = bytes[i] & 0xFF;
            result[i * 2] = table[b >>> 4];
            result[i * 2 + 1] = table[b & 0x0F];
        }
        return new String(result);
    }

    private static ProvenResult<byte[]> decodePure(String hex) {
        byte[] result = new byte[hex.length() / 2];
        for (int i = 0; i < result.length; i++) {
            int hi = Character.digit(hex.charAt(i * 2), 16);
            int lo = Character.digit(hex.charAt(i * 2 + 1), 16);
            if (hi < 0 || lo < 0) {
                return ProvenResult.err(ProvenStatus.ERR_INVALID_ARGUMENT,
                    "Invalid hex character at position " + (i * 2));
            }
            result[i] = (byte) ((hi << 4) | lo);
        }
        return ProvenResult.ok(result);
    }

    private static boolean constantTimeEqualsPure(String a, String b) {
        int result = 0;
        for (int i = 0; i < a.length(); i++) {
            char ca = Character.toLowerCase(a.charAt(i));
            char cb = Character.toLowerCase(b.charAt(i));
            result |= ca ^ cb;
        }
        return result == 0;
    }
}
